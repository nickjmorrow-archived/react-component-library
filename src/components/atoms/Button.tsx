import * as React from "react";
import styled from "styled-components";
import { colors, transitions, borderRadius, boxShadow } from "styleConstants";
import { Typography } from "./Typography";
import { Link } from "react-router-dom";

interface IDisplayProps {
  color: ButtonVariant;
  backgroundColor: ButtonVariant;
  showBoxShadow?: boolean;
  useMargin?: boolean;
  style?: React.CSSProperties;
}

export type ButtonVariant =
  | "primary"
  | "secondary"
  | "cancel"
  | "white"
  | "transparent"
  | "google";

export type IButtonProps = {
  color?: ButtonVariant;
  variant?: ButtonVariant;
  route?: string;
  children: React.ReactNode;
  onClick?(): void;
} & Partial<IDisplayProps>;

export const Button: React.SFC<IButtonProps> = ({
  children,
  variant = "primary" as ButtonVariant,
  route,
  showBoxShadow = true,
  useMargin = true,
  color = "white" as ButtonVariant,
  style,
  onClick: handleClick = () => {
    return;
  }
}) => {
  const formattedChildren =
    typeof children === "string" ? (
      <Typography color="inherit" variant="button">
        {children}
      </Typography>
    ) : (
      children
    );
  const button = (
    <StyledButton
      color={color}
      backgroundColor={variant}
      showBoxShadow={showBoxShadow}
      useMargin={useMargin}
      onClick={handleClick}
      style={style}>
      {formattedChildren}
    </StyledButton>
  );
  return route ? (
    <Link to={route} style={{ textDecoration: "none" }}>
      {button}
    </Link>
  ) : (
    button
  );
};

const StyledButton = styled("button")<IDisplayProps>`
  color: ${props => getColor(props.color)};
  background-color: ${props => getColor(props.backgroundColor)};
  border-radius: ${borderRadius.default};
  padding: 8px 16px;
  height: 36px;
  display: flex;
  border: none;
  justify-content: center;
  align-items: center;
  margin: ${props => (props.useMargin ? "4px" : "0px")};
  cursor: pointer;
  outline: none;
  box-shadow: ${props => props.showBoxShadow && boxShadow.default};
  width: max-content;
  &:hover,
  &:focus {
    background-color: ${props => getColorHover(props.backgroundColor)};
    color: ${props => getColorHover(props.color)};
    transition: all ${transitions.fast};
  }
  &:active {
    background-color: ${props => getColorActive(props.backgroundColor)};
    color: ${props => getColorActive(props.color)};
    transition: all ${transitions.fast};
  }
`;

// helpers
const getColorHover = (variant: ButtonVariant) => {
  switch (variant) {
    case "primary":
      return colors.primaryLight;
    case "secondary":
      return colors.secondaryLight;
    case "cancel":
      return colors.tertiaryLight;
    case "white":
      return colors.white;
    case "transparent":
      return colors.transparent;
    case "google":
      return colors.googleRedLight;
  }
};

const getColorActive = (variant: ButtonVariant) => {
  switch (variant) {
    case "primary":
      return colors.primaryDark;
    case "secondary":
      return colors.secondaryDark;
    case "cancel":
      return colors.tertiaryDark;
    case "white":
      return colors.white;
    case "transparent":
      return colors.transparent;
    case "google":
      return colors.googleRedDark;
  }
};

const getColor = (variant: ButtonVariant) => {
  switch (variant) {
    case "white":
      return colors.white;
    case "primary":
      return colors.primary;
    case "secondary":
      return colors.secondary;
    case "cancel":
      return colors.tertiaryDark;
    case "transparent":
      return colors.transparent;
    case "google":
      return colors.googleRed;
  }
};
