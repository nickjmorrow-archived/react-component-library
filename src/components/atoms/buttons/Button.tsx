import * as React from "react";
import styled from "styled-components";
import { ThemeContext } from "~/styleConstants";
import { StyleConstant } from "~/typeUtilities";
import { ColorVariant, ColorSet, StyleVariant, ColorState } from "../types";
import { Typography } from "../typography/Typography";
import { getBackgroundColorState, getBorderColorState } from "./buttonServices";
import { ButtonColorVariant } from "./types";
import PulseLoader from "react-spinners/PulseLoader";
import { Fade } from "../../animations";
import { Link } from "react-router-dom";
import { getColor } from "../atomServices";

interface IDisplayProps {
  showBoxShadow?: boolean;
  useMargin?: boolean;
  style?: React.CSSProperties;
  backgroundColorState: ColorState;
  borderColorState: ColorState;
  boxShadow: StyleConstant<"boxShadow">;
  border: StyleConstant<"border">;
  transition: string;
  isFullWidth: boolean;
  styleVariant: StyleVariant;
  spacing: StyleConstant<"spacing">;
  width: number;
  height: number;
}

type IButtonProps = {
  textColorVariant?: ColorVariant;
  colorVariant?: ButtonColorVariant;
  styleVariant?: StyleVariant;
  route?: string;
  children?: React.ReactNode;
  colorSet?: Partial<ColorSet>;
  isFullWidth?: boolean;
  isLoading?: boolean;
  link?: string;
  onClick?(): void;
} & Partial<IDisplayProps> &
  Partial<ColorSet>;

const ButtonInternal: React.SFC<IButtonProps> = ({
  children,
  colorVariant = "core",
  textColorVariant = "primaryLight",
  styleVariant = "primary",
  showBoxShadow = true,
  useMargin = true,
  isFullWidth = false,
  isLoading,
  link,
  colorSet = {} as ColorSet,
  onClick: handleClick = () => {
    return;
  }
}) => {
  const typographyColorSet =
    styleVariant === "primary"
      ? colorSet
      : {
          color: colorSet.backgroundColor,
          colorHover: colorSet.backgroundColorHover,
          colorActive: colorSet.backgroundColorActive
        };
  const formattedChildren =
    typeof children === "string" ? (
      <Typography
        colorVariant={textColorVariant}
        sizeVariant={2}
        weightVariant={2}
        isInteractive={true}
        colorSet={typographyColorSet}
        style={{ textTransform: "uppercase" }}>
        {children}
      </Typography>
    ) : (
      children
    );

  const { colors, transitions, boxShadow, spacing, border } = React.useContext(
    ThemeContext
  );

  const innerWrapperRef = React.useRef<HTMLDivElement>(null);
  const [width, setWidth] = React.useState(0);
  const [height, setHeight] = React.useState(0);

  if (innerWrapperRef && innerWrapperRef.current) {
    if (innerWrapperRef.current.clientWidth > width) {
      setWidth(innerWrapperRef.current.clientWidth);
    }

    if (innerWrapperRef.current.clientHeight > height) {
      setHeight(innerWrapperRef.current.clientHeight);
    }
  }

  const backgroundColorState = getBackgroundColorState(
    colors,
    colorVariant,
    colorSet,
    styleVariant
  );

  const borderColorState = getBorderColorState(
    colors,
    colorVariant,
    colorSet,
    styleVariant
  );

  const loadingFade = (
    <>
      <Fade
        in={isLoading as boolean}
        style={{ position: "absolute" }}
        transitionVariant={"medium"}>
        <PulseLoader
          color={getColor(colors, textColorVariant)}
          size={8}
          sizeUnit={"px"}
        />
      </Fade>
      <Fade in={!isLoading} transitionVariant={"medium"}>
        {formattedChildren}
      </Fade>
    </>
  );

  const content = (
    <StyledButton
      isFullWidth={isFullWidth}
      width={width}
      height={height}
      backgroundColorState={backgroundColorState}
      borderColorState={borderColorState}
      showBoxShadow={showBoxShadow}
      useMargin={useMargin}
      onClick={handleClick}
      transition={transitions.medium}
      border={border}
      styleVariant={styleVariant}
      boxShadow={boxShadow}
      spacing={spacing}>
      <InnerWrapper width={width} height={height} ref={innerWrapperRef}>
        {isLoading != undefined ? loadingFade : formattedChildren}
      </InnerWrapper>
    </StyledButton>
  );

  return link ? (
    <Link to={link} style={{ textDecoration: "none" }}>
      {content}
    </Link>
  ) : (
    content
  );
};

const InnerWrapper = styled("div")<{ width: number; height: number }>`
  min-width: ${p => p.width}px;
  min-height: ${p => p.height}px;
  display: flex;
  align-items: center;
  justify-content: center;
`;

const StyledButton = styled("button")<
  IDisplayProps & Partial<ColorSet & React.HTMLProps<HTMLButtonElement>>
>`
  color: ${p => (p.styleVariant === "secondary" ? p.backgroundColor : p.color)};
  background-color: ${p => p.backgroundColorState.normal};
  border-radius: ${props => props.border.borderRadius.br1};
  border: ${p => `${p.border.borderStyle.bs2} ${p.borderColorState.normal}`};
  padding: ${p => p.spacing.ss3 + " " + p.spacing.ss4};
  display: flex;
  justify-content: center;
  align-items: center;
  margin: ${props => (props.useMargin ? props.spacing.ss4 : 0)};
  cursor: pointer;
  outline: none;
  word-wrap: no-wrap;
  box-shadow: ${props => props.showBoxShadow && props.boxShadow.bs2};
  min-width: ${p => p.width}px;
  min-height: ${p => p.height}px;
  height: max-content;
  ${p => (p.isFullWidth ? "width: 100%" : "")}
  transition: box-shadow ${p => p.transition},
    background-color ${p => p.transition}, border-color ${p => p.transition};
  &:hover {
    border-color: ${p => p.borderColorState.hover};
    background-color: ${p => p.backgroundColorState.hover};
    color: ${p =>
      p.styleVariant === "secondary" ? p.backgroundColorHover : p.colorHover};
    box-shadow: ${props => props.showBoxShadow && props.boxShadow.bs1};
    transition: all ${props => props.transition} ease-in-out;
  }
  &:active {
    border-color: ${p => p.borderColorState.active};
    background-color: ${p => p.backgroundColorState.active};
    color: ${props => props.colorActive};
    transition: all ${props => props.transition} ease-in-out;
  }
`;

export const Button = React.memo(ButtonInternal);
