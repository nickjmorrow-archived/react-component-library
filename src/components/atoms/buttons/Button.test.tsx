import * as React from "react";
import { Button } from "./Button";
import * as TestRenderer from "react-test-renderer/shallow";

const renderer = TestRenderer.createRenderer();
describe("Button rendering", () => {
  it("default", () => {
    const tree = renderer.render(<Button />);
    expect(tree).toMatchSnapshot();
  });
  it("primary", () => {
    const tree = renderer.render(<Button styleVariant={"primary"} />);
    expect(tree).toMatchSnapshot();
  });
  it("secondary", () => {
    const tree = renderer.render(<Button styleVariant={"secondary"} />);
    expect(tree).toMatchSnapshot();
  });
  it("tertiary", () => {
    const tree = renderer.render(<Button styleVariant={"tertiary"} />);
    expect(tree).toMatchSnapshot();
  });
});
