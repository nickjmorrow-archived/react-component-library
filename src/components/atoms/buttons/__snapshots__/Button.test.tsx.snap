// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button rendering default 1`] = `
<UNDEFINED
  backgroundColorState={
    Object {
      "active": "hsl(220, 55%, 40%)",
      "hover": "hsl(220, 55%, 60%)",
      "normal": "hsl(220, 55%, 50%)",
    }
  }
  border={
    Object {
      "borderRadius": Object {
        "br1": "6px",
        "br2": "6px",
      },
      "borderStyle": Object {
        "bs1": "1px solid",
        "bs2": "2px solid",
        "bs3": "5px solid",
      },
    }
  }
  borderColorState={
    Object {
      "active": "hsl(220, 55%, 40%)",
      "hover": "hsl(220, 55%, 60%)",
      "normal": "hsl(220, 55%, 50%)",
    }
  }
  boxShadow={
    Object {
      "bs1": "0 1px 3px hsla(0, 0%, 0%, 0.2), 0 1px 2px hsla(0, 0%, 0%, 0.2)",
      "bs2": "0 3px 6px hsla(0, 0%, 0%, 0.2), 0 2px 4px hsla(0, 0%, 0%, 0.2)",
      "bs3": "0 10px 20px hsla(0, 0%, 0%, 0.2), 0 3px 6px hsla(0, 0%, 0%, 0.2)",
      "bs4": "0 15px 25px hsla(0, 0%, 0%, 0.2), 0 5px 10px hsla(0, 0%, 0%, 0.2)",
      "bs5": "0 20px 40px hsla(0, 0%, 0%, 0.2)",
    }
  }
  height={0}
  isFullWidth={false}
  onClick={[Function]}
  showBoxShadow={true}
  spacing={
    Object {
      "ss1": "4px",
      "ss12": "48px",
      "ss128": "512px",
      "ss16": "64px",
      "ss160": "640px",
      "ss192": "768px",
      "ss2": "8px",
      "ss24": "96px",
      "ss3": "12px",
      "ss32": "128px",
      "ss4": "16px",
      "ss48": "192px",
      "ss6": "24px",
      "ss64": "256px",
      "ss8": "32px",
      "ss96": "384px",
    }
  }
  styleVariant="primary"
  transition="350ms cubic-bezier(0.645, 0.045, 0.355, 1.000)"
  useMargin={true}
  width={0}
>
  <UNDEFINED
    height={0}
    width={0}
  />
</UNDEFINED>
`;

exports[`Button rendering primary 1`] = `
<UNDEFINED
  backgroundColorState={
    Object {
      "active": "hsl(220, 55%, 40%)",
      "hover": "hsl(220, 55%, 60%)",
      "normal": "hsl(220, 55%, 50%)",
    }
  }
  border={
    Object {
      "borderRadius": Object {
        "br1": "6px",
        "br2": "6px",
      },
      "borderStyle": Object {
        "bs1": "1px solid",
        "bs2": "2px solid",
        "bs3": "5px solid",
      },
    }
  }
  borderColorState={
    Object {
      "active": "hsl(220, 55%, 40%)",
      "hover": "hsl(220, 55%, 60%)",
      "normal": "hsl(220, 55%, 50%)",
    }
  }
  boxShadow={
    Object {
      "bs1": "0 1px 3px hsla(0, 0%, 0%, 0.2), 0 1px 2px hsla(0, 0%, 0%, 0.2)",
      "bs2": "0 3px 6px hsla(0, 0%, 0%, 0.2), 0 2px 4px hsla(0, 0%, 0%, 0.2)",
      "bs3": "0 10px 20px hsla(0, 0%, 0%, 0.2), 0 3px 6px hsla(0, 0%, 0%, 0.2)",
      "bs4": "0 15px 25px hsla(0, 0%, 0%, 0.2), 0 5px 10px hsla(0, 0%, 0%, 0.2)",
      "bs5": "0 20px 40px hsla(0, 0%, 0%, 0.2)",
    }
  }
  height={0}
  isFullWidth={false}
  onClick={[Function]}
  showBoxShadow={true}
  spacing={
    Object {
      "ss1": "4px",
      "ss12": "48px",
      "ss128": "512px",
      "ss16": "64px",
      "ss160": "640px",
      "ss192": "768px",
      "ss2": "8px",
      "ss24": "96px",
      "ss3": "12px",
      "ss32": "128px",
      "ss4": "16px",
      "ss48": "192px",
      "ss6": "24px",
      "ss64": "256px",
      "ss8": "32px",
      "ss96": "384px",
    }
  }
  styleVariant="primary"
  transition="350ms cubic-bezier(0.645, 0.045, 0.355, 1.000)"
  useMargin={true}
  width={0}
>
  <UNDEFINED
    height={0}
    width={0}
  />
</UNDEFINED>
`;

exports[`Button rendering secondary 1`] = `
<UNDEFINED
  backgroundColorState={
    Object {
      "active": "transparent",
      "hover": "transparent",
      "normal": "transparent",
    }
  }
  border={
    Object {
      "borderRadius": Object {
        "br1": "6px",
        "br2": "6px",
      },
      "borderStyle": Object {
        "bs1": "1px solid",
        "bs2": "2px solid",
        "bs3": "5px solid",
      },
    }
  }
  borderColorState={
    Object {
      "active": "hsl(220, 55%, 40%)",
      "hover": "hsl(220, 55%, 60%)",
      "normal": "hsl(220, 55%, 50%)",
    }
  }
  boxShadow={
    Object {
      "bs1": "0 1px 3px hsla(0, 0%, 0%, 0.2), 0 1px 2px hsla(0, 0%, 0%, 0.2)",
      "bs2": "0 3px 6px hsla(0, 0%, 0%, 0.2), 0 2px 4px hsla(0, 0%, 0%, 0.2)",
      "bs3": "0 10px 20px hsla(0, 0%, 0%, 0.2), 0 3px 6px hsla(0, 0%, 0%, 0.2)",
      "bs4": "0 15px 25px hsla(0, 0%, 0%, 0.2), 0 5px 10px hsla(0, 0%, 0%, 0.2)",
      "bs5": "0 20px 40px hsla(0, 0%, 0%, 0.2)",
    }
  }
  height={0}
  isFullWidth={false}
  onClick={[Function]}
  showBoxShadow={true}
  spacing={
    Object {
      "ss1": "4px",
      "ss12": "48px",
      "ss128": "512px",
      "ss16": "64px",
      "ss160": "640px",
      "ss192": "768px",
      "ss2": "8px",
      "ss24": "96px",
      "ss3": "12px",
      "ss32": "128px",
      "ss4": "16px",
      "ss48": "192px",
      "ss6": "24px",
      "ss64": "256px",
      "ss8": "32px",
      "ss96": "384px",
    }
  }
  styleVariant="secondary"
  transition="350ms cubic-bezier(0.645, 0.045, 0.355, 1.000)"
  useMargin={true}
  width={0}
>
  <UNDEFINED
    height={0}
    width={0}
  />
</UNDEFINED>
`;

exports[`Button rendering tertiary 1`] = `
<UNDEFINED
  backgroundColorState={
    Object {
      "active": "transparent",
      "hover": "transparent",
      "normal": "transparent",
    }
  }
  border={
    Object {
      "borderRadius": Object {
        "br1": "6px",
        "br2": "6px",
      },
      "borderStyle": Object {
        "bs1": "1px solid",
        "bs2": "2px solid",
        "bs3": "5px solid",
      },
    }
  }
  borderColorState={
    Object {
      "active": "transparent",
      "hover": "transparent",
      "normal": "transparent",
    }
  }
  boxShadow={
    Object {
      "bs1": "0 1px 3px hsla(0, 0%, 0%, 0.2), 0 1px 2px hsla(0, 0%, 0%, 0.2)",
      "bs2": "0 3px 6px hsla(0, 0%, 0%, 0.2), 0 2px 4px hsla(0, 0%, 0%, 0.2)",
      "bs3": "0 10px 20px hsla(0, 0%, 0%, 0.2), 0 3px 6px hsla(0, 0%, 0%, 0.2)",
      "bs4": "0 15px 25px hsla(0, 0%, 0%, 0.2), 0 5px 10px hsla(0, 0%, 0%, 0.2)",
      "bs5": "0 20px 40px hsla(0, 0%, 0%, 0.2)",
    }
  }
  height={0}
  isFullWidth={false}
  onClick={[Function]}
  showBoxShadow={true}
  spacing={
    Object {
      "ss1": "4px",
      "ss12": "48px",
      "ss128": "512px",
      "ss16": "64px",
      "ss160": "640px",
      "ss192": "768px",
      "ss2": "8px",
      "ss24": "96px",
      "ss3": "12px",
      "ss32": "128px",
      "ss4": "16px",
      "ss48": "192px",
      "ss6": "24px",
      "ss64": "256px",
      "ss8": "32px",
      "ss96": "384px",
    }
  }
  styleVariant="tertiary"
  transition="350ms cubic-bezier(0.645, 0.045, 0.355, 1.000)"
  useMargin={true}
  width={0}
>
  <UNDEFINED
    height={0}
    width={0}
  />
</UNDEFINED>
`;
